docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+We can enable the use of Cluster API in these use-cases by updating our providers to perform in-place upgrades.
+
+
+## User facing changes
+<!--
+This section MUST describe any user-facing changes that this feature brings, if
+any. If an API change is required, the affected endpoints MUST be mentioned. If
+the output of any k8s command changes, the difference MUST be mentioned, with a
+clear example of "before" and "after".
+-->
+
+Users will be able to perform in-place upgrades per machine basis by running:
+```sh
+kubectl annotate machine <machine-name> k8sd.io/in-place-upgrade={upgrade-option}
+```
+
+Users can also perform in-place upgrades on the entire cluster by running:
+```sh
+kubectl annotate cluster <cluster-name> k8sd.io/in-place-upgrade={upgrade-option}
+```
+This would upgrade machines belonging to `<cluster-name>` one by one.
+
+`{upgrade-option}` can be one of:
+* `channel=<channel>` which would refresh the machine to the provided channel e.g. `channel=1.31-classic/stable`
+* `revision=<revision>` which would refresh the machine to the provided revision e.g. `revision=640`
+* `localPath=<path-to-file>` which would refresh the machine to the provided local `*.snap` file e.g. `localPath=/path/to/k8s.snap`
+
+## Alternative solutions
+<!--
+This section SHOULD list any possible alternative solutions that have been or
+should be considered. If required, add more details about why these alternative
+solutions were discarded.
+-->
+
+We could alternatively use the `version` fields defined in `ControlPlane` and `MachineDeployment` manifests instead of annotations which could be a better/more native user experience.
+
+However at the time of writing CAPI does not have support for changing upgrade strategies which means changes to the `version` fields trigger a rolling update.
+
+This behaviour can be adjusted on `ControlPlane` objects as our provider has more/full control but can not be easily adjusted on `MachineDeployment` objects which causes issues for worker nodes.
+
+Switching to using the `version` field should take place when upstream implements support for different upgrade strategies.
+
+## Out of scope
+<!--
+This section MUST reference any work that is out of scope for this proposal.
+Out of scope items are typically unknowns that we do not yet have a clear idea
+of how to solve, so we explicitly do not tackle them until we have more
+information.
+
+This section is very useful to help guide the implementation details section
+below, or serve as reference for future proposals.
+-->
+
+none
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+### `POST /x/capi/snap-refresh`
```

Comment:
The `snap-refresh` endpoint is not really CAPI specific. Right now, we will only use it there but who knows...

Maybe `/x/snap/refresh` makes more sense?

Reply:
Wouldn't we need to create a CAPI specific endpoint like we've done for `x/capi/remove-node` to do CAPI specific auth?

Reply:
true, good point. Ignore what I wrote.